service_test.go

```diff
@@ -4,29 +4,70 @@ import (
 	"fmt"
 	"github.com/stretchr/testify/assert"
 	"testing"
-	"time"
 )
 
-func Test_GivenTickerServiceIsStarted_WhenNewMeasurementComes_ThenReadingIsPassedToSaveChannel(t *testing.T) {
-	dao := &DataAccessObject{make(map[int][]DeviceReading)}
-	s := Service{}
-	s.init(dao)
-	s.run()
+type mockDao struct {
+	returnValue int
+	returnErr   error
+	calledTimes int
+}
+
+func (m *mockDao) AddDevice(device *DevicePayload) (int, error) {
+	m.calledTimes++
+	m.returnValue++
+	return m.returnValue, m.returnErr
+}
 
-	input := &RawInput{
-		Id:       "0",
-		Name:     "TestDevice",
-		Interval: "1000",
+func Test_CorrectDevice_ServiceSavesNewDevice(t *testing.T) {
+	device := &DevicePayload{
+		Value:    10.23,
+		Name:     "Thermostat",
+		Interval: 1000,
 	}
-	dev, _ := s.createDevice(input)
-	m := func(n int) string { return "10C" }
-	_ = s.startDevice(dev, m)
+	dao := &mockDao{returnValue: 1}
+
+	out := Service{Dao: dao}
+
+	result, err := out.AddDevice(device)
+
+	assert.NoError(t, err)
+	assert.Equal(t, 1, dao.calledTimes)
+	assert.NotNil(t, result)
+}
+
+func Test_CorrectDeviceAndDaoFails_ServiceFails(t *testing.T) {
+	dao := &mockDao{returnErr: fmt.Errorf("test error")}
+	out := Service{Dao: dao}
 
-	time.Sleep(1 * time.Second)
-	s.stop()
+	_, err := out.AddDevice(&DevicePayload{
+		Value:    10.23,
+		Name:     "Thermostat",
+		Interval: 1000,
+	})
 
-	expected := "10C"
-	result := fmt.Sprintf("%v", s.DevicesReadings.data[0][0].Value)
-	fmt.Println(expected)
-	assert.Equal(t, expected, result)
+	assert.Error(t, err)
+}
+
+func Test_GivenIntervalValueBelowZeroOrEqualToZero_ServiceFails(t *testing.T) {
+	out := Service{Dao: &mockDao{}}
+
+	_, err1 := out.AddDevice(&DevicePayload{Interval: -1})
+	_, err2 := out.AddDevice(&DevicePayload{Interval: 0})
+	assert.Error(t, err1)
+	assert.Error(t, err2)
+}
+
+func Test_CorrectPayload_DaoFillsOutId_ServiceDefaultsInterval(t *testing.T) {
```

Comment:
The test name tells about two things:
- filling id returned by dao
- setting default interval

Usually you want to test one small thing at a time.